src/join.sol

```diff
@@ -427,3 +428,74 @@ contract AuthGemJoin is LibNote {
         require(gem.transfer(usr, wad), "AuthGemJoin/failed-transfer");
     }
 }
+
+// GemJoin7
+// For an upgradable token (like USDT)
+//  If the token is deprecated changing the implementation behind, this prevents joins
+//   and exits until the implementation is reviewed and approved by governance.
+
+interface GemLike7 {
+    function decimals() external view returns (uint);
+    function transfer(address,uint) external;
+    function transferFrom(address,address,uint) external;
+    function balanceOf(address) external view returns (uint);
+    function upgradedAddress() external view returns (address);
+    function setImplementation(address,uint) external;
+    function adjustFee(uint) external;
+}
+
+contract GemJoin7 is LibNote {
+    mapping (address => uint) public wards;
+    function rely(address usr) external note auth { wards[usr] = 1; }
+    function deny(address usr) external note auth { wards[usr] = 0; }
+    modifier auth { require(wards[msg.sender] == 1); _; }
+
+    VatLike  public vat;
+    bytes32  public ilk;
+    GemLike7 public gem;
+    uint     public dec;
+    uint     public live; // Access flag
+
+    mapping (address => uint256) public implementations;
+
+    constructor(address vat_, bytes32 ilk_, address gem_) public {
+        gem = GemLike7(gem_);
+        dec = gem.decimals();
+        require(dec < 18, "GemJoin7/decimals-18-or-higher");
+        wards[msg.sender] = 1;
+        live = 1;
+        vat = VatLike(vat_);
+        ilk = ilk_;
+        setImplementation(address(gem.upgradedAddress()), 1);
+    }
+
+    function cage() external note auth {
+        live = 0;
+    }
+
+    function setImplementation(address implementation, uint256 permitted) public auth note {
+        implementations[implementation] = permitted; // 1 live, 0 disable
+    }
+
+    function mul(uint x, uint y) internal pure returns (uint z) {
+        require(y == 0 || (z = x * y) / y == x, "GemJoin7/overflow");
+    }
+
+    function join(address urn, uint wad) public note {
+        require(live == 1, "GemJoin7/not-live");
+        require(implementations[gem.upgradedAddress()] == 1, "GemJoin7/implementation-invalid");
+        uint bal = gem.balanceOf(address(this));
+        gem.transferFrom(msg.sender, address(this), wad);
+        uint wadt = mul(gem.balanceOf(address(this)) - bal, 10 ** (18 - dec));
```

Comment:
I'd be extra safe and would use `sub` instead of `-` for the balance difference.

Reply:
+1

Reply:
fixed in latest commit